geo/src/algorithm/coordinate_position.rs

```diff
@@ -0,0 +1,749 @@
+use crate::algorithm::{
+    bounding_rect::BoundingRect, dimensions::HasDimensions, intersects::Intersects,
+    kernels::HasKernel,
+};
+use crate::{
+    Coordinate, Geometry, GeometryCollection, Line, LineString, MultiLineString, MultiPoint,
+    MultiPolygon, Point, Polygon, Rect, Triangle,
+};
+
+/// The position of a `Coordinate` relative to a `Geometry`
+#[derive(PartialEq, Clone, Debug)]
+pub enum CoordPos {
+    OnBoundary,
+    Inside,
+    Outside,
+}
+
+/// Determine the position of a `Coordinate` relative to a geometry.
+///
+/// # Examples
+///
+/// ```rust
+/// use geo::{polygon, Coordinate};
+/// use geo::coordinate_position::{CoordinatePosition, CoordPos};
+///
+/// let square_poly = polygon![(x: 0.0, y: 0.0), (x: 2.0, y: 0.0), (x: 2.0, y: 2.0), (x: 0.0, y: 2.0), (x: 0.0, y: 0.0)];
+///
+/// let inside_coord = Coordinate { x: 1.0, y: 1.0 };
+/// assert_eq!(square_poly.coordinate_position(&inside_coord), CoordPos::Inside);
+///
+/// let boundary_coord = Coordinate { x: 0.0, y: 1.0 };
+/// assert_eq!(square_poly.coordinate_position(&boundary_coord), CoordPos::OnBoundary);
+///
+/// let outside_coord = Coordinate { x: 5.0, y: 5.0 };
+/// assert_eq!(square_poly.coordinate_position(&outside_coord), CoordPos::Outside);
+/// ```
+pub trait CoordinatePosition<T>
```

Comment:
@michaelkirk Do you prefer the `CoordinatePosition<T>` in favor of `CoordinatePosition` with a associate type `Scalar: CoordinateType` ?  My personal pref. is the latter;  optional comment.

Reply:
I believe I've done this here: a60dc27 - can you take a look @rmanoka?

Despite reading a couple articles on this, I only kind of theoretically understand why this distinction is good, and still don't have a very intuitive grasp for when to apply it.

Is your rule of thumb to prefer associated types and only use full-blown generic type params when associated types are insufficient?

Reply:
I'll plan to merge this in a couple days if I don't hear otherwise.

Reply:
@michaelkirk Sorry for the late response;  the changes look great!   My pref. is mostly similar to `Iterator` vs `AddAsign` traits in std.  I go for an associated type unless the same type could implement the trait with two possible param.  Here, we're implementing the trait `CoordinatePosition` with only `T = Ty` for `Line<Ty>`, so prefer it to be associated than make the trait `CoordinatePosition<T>`.  

May be in the future, our scalar types could have additional sophistication that tells us when `T1` can be converted to `T2` without loss in precision, and we can make some of our traits like `CoordinatePosition`, `ForEach` into generics.